src/modulefinder/sentry_modulefinder_linux.c

```diff
@@ -419,13 +475,53 @@ sentry__procmaps_module_to_value(const sentry_module_t *module)
     const sentry_mapped_region_t *first_mapping = &module->mappings[0];
     const sentry_mapped_region_t *last_mapping
         = &module->mappings[module->num_mappings - 1];
+    uint64_t module_size
+        = last_mapping->addr + last_mapping->size - first_mapping->addr;
+
     sentry_value_set_by_key(
         mod_val, "image_addr", sentry__value_new_addr(first_mapping->addr));
-    sentry_value_set_by_key(mod_val, "image_size",
-        sentry_value_new_int32(
-            last_mapping->addr + last_mapping->size - first_mapping->addr));
+    sentry_value_set_by_key(
+        mod_val, "image_size", sentry_value_new_int32(module_size));
+
+    // At least on the android API-16, x86 simulator, the linker apparently
+    // does not load the complete file into memory. Or at least, the section
+    // headers which are located at the end of the file are not loaded, and
+    // we would be poking into invalid memory. To be safe, we mmap the
+    // complete file from disk, so we have the on-disk layout, and are
+    // independent of how the runtime linker would load or re-order any
+    // sections. The exception here is the linux-gate, which is not an
+    // actual file on disk, so we actually poke at its memory.
+    if (sentry__slice_eq(module->file, LINUX_GATE)) {
+        sentry__procmaps_read_ids_from_elf(mod_val, module);
+    } else {
+        char *filename = sentry__slice_to_owned(module->file);
+        sentry_mmap_t mm;
+        if (!sentry__mmap_file(&mm, filename)) {
+            sentry_free(filename);
+            sentry_value_decref(mod_val);
+            return sentry_value_new_null();
+        }
+        sentry_free(filename);
+
+        uint64_t needed_mmap_size = module->offset_in_inode + module_size;
+        if (mm.len < needed_mmap_size) {
```

Comment:
For me it seems that mappings are larger than the file they are mapping. This results in them failing as this is always true.

For example:

problem file: `libicui18n.so.69.1`

```bash
    # ls -l /usr/lib/libicui18n.so.69.1
    -rwxr-xr-x    1 root     root       2320048 Dec 23 06:07 /usr/lib/libicui18n.so.69.1
```

2,320,048 bytes = 0x23 66B0

```bash
    # cat /proc/688/maps | grep icui
    99255000-9948c000 r--p 00000000 b3:06 1853       /usr/lib/libicui18n.so.69.1 (0x  23 7000 = 2,322,432)
    b2481000-b26af000 r-xp 00000000 b3:06 1853       /usr/lib/libicui18n.so.69.1 (0x  22 E000 = 2,285,568)
    b26af000-b26be000 ---p 0022e000 b3:06 1853       /usr/lib/libicui18n.so.69.1
    b26be000-b26c6000 r--p 0022d000 b3:06 1853       /usr/lib/libicui18n.so.69.1 (0x     8000 =    32,768)
    b26c6000-b26c8000 rw-p 00235000 b3:06 1853       /usr/lib/libicui18n.so.69.1 (0x     2000 =     8,192)
```

the `99255000` mapping is too large for the file size
and the `b26c6000` mapping also overruns the file

needed size is 0x24 7000 but the file size is only 0x23 66b0

I replaced the code with this, and it seems to work for me, however I'm not sure if this is valid or not.

```cpp
        // uint64_t needed_mmap_size = module->offset_in_inode + module_size;
        // if (mm.len < needed_mmap_size) {
        //     sentry__mmap_close(&mm);
        //     sentry_value_decref(mod_val);
        //     return sentry_value_new_null();
        // }

        sentry_module_t mmapped_module;
        memset(&mmapped_module, 0, sizeof(sentry_module_t));
        mmapped_module.mappings[0].size = mm.len; // module_size;
```

Reply:
👍🏻 That sounds very good!
I think the mapping can be either bigger because the loader is rounding up to page sizes. Or smaller if the loader just does not map some "unneeded" sections. Using the on-disk size should work there.

Thanks for taking a look!